### ISSUE TITLE
etcd: Use protobuf instead of JSON as the storage format

### ISSUE THREAD

**Issue by asdine** [heetch/regula#25]
The UI will encourage people to make small iterations and hence create more and more versions. In order to give us more room regarding storage size and encoding/decoding speed of database data, we will switch to Protobuf for storing data in etcd.

Here is a list of Protobuf message that can be used to translate JSON rulesets into:

```protobuf
    message Ruleset {
        repeated Rule rules = 1;
        string type = 2;
    }

     message Rule {
        Expr expr = 1;
        Value result = 2;
    }

     message Expr {
        oneof expr {
            Value value = 1;
            Param param = 2;
            Operator operator = 3;
        }
    }

     message Value {
        string kind = 1;
        string type = 2;
        string data = 3;
    }

     message Param {
        string kind = 1;
        string type = 2;
        string name = 3;
    }

     message Operator {
        string kind = 1;
        repeated Expr operands = 2;
    }
```

Signatures:

```protobuf
    message Signature {
        string returnType = 1;
        map<string, string> paramTypes = 2;
    }
```

**Comment by jhchabran** [comment]
@asdine what is our current usage in terms of storage size? 

If we have still a large margin, there's no pressing need for applying this change and thus we can schedule it after having done all the most important issues, such as API changes, etc ... Also, it could be easier to just have JSON while we're debugging. 

Of course, we could do this once we've dealt with all "mandatory" changes that would bring us closer to the release.

Do you see any reason to not take this approach?

**Comment by asdine** [comment]
Is this about scheduling or about the change itself?

**Comment by YassineE** [comment]
@jhchabran / @asdine what is the state of this discussion?

**Comment by asdine** [comment]
This is not a priority for now. We will try to ship it for this version as much as possible, but if the scheduling doesn't allow it, we will ship it during the next version, as an important feature.

**Comment by YassineE** [comment]
Should I jump on another issue in the backlog then?

**Comment by YassineE** [comment]
Fixed in #86.
